### DISCUSSION TITLE
Production deployment

### DISCUSSION THREAD

**Discussion by vatsagravat** [NRules/NRules#272] [Q&A]
How to deploy rule changes in a production environment? Is it useful for dynamic rules? Is it a good approach if the rules are changed frequently? If not, any suggestions.

**Comment by snikolayev** [comment]
NRules is a library, so to use it in a production environment, there must be some sort of a runtime component (e.g. a service) that hosts the engine and loads the rules. It's a good idea to compile rules into separate assemblies, such that you can load those assemblies on service start, dynamically, compile the rules into a rete graph (session factory), and then use it throughout the lifetime of the service.
Then upgrading rules would normally mean shutting down the service, deploying new rules and restarting the service. There are many deployment patterns that exist that make this more robust, e.g. scaling the service out and processing your transactions on multiple instances. Then, you can spin up new services with new rules and gracefully move traffic from the old instances to the new ones and tear down the old instances w/o any downtime. 
This is just one possibility. The main point is that NRules is a library that implements the rules engine. It's not a rules execution environment nor a full BRE ecosystem. So, it does not put any constraints on how you use it.
